### PR TITLE
[design] Make it possible to pass AlgorithmIdentifier parameter data

### PR DESCRIPTION
**Pull request by levitte**
This design is to allow passing AlgorithmIdentifier parameter data to and
from any cryptograpfic operation, with convenience functions for them all,
not just for symmetric ciphers.

This is crucial to support CMS, among others.

Ref: openssl/project#172

### CODE REVIEW

File: doc/designs/passing-algorithmidentifier-parameters.md [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+Note that all might not need to be added immediately, depending on if they
+are considered useful or not.  For future proofing, however, they should
+probably all be added.
+
+Requirements on the providers
+-----------------------------
+
+Providers that implement ciphers or any operation that uses asymmetric keys
+will have to implement support for passing AlgorithmIdentifier parameter
+data, and will to process that data in whatever manner that's necessary to
```
Comment (mattcaswell) [inline doc/designs/passing-algorithmidentifier-parameters.md:74]: "and will to process"....doesn't quite make sense

File: doc/designs/passing-algorithmidentifier-parameters.md [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+meet the standards for that operation.
```
Comment (mattcaswell) [inline doc/designs/passing-algorithmidentifier-parameters.md:75]: Can we have some examples of things that might be necessary? I haven't quite got my head around how this is used yet.

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: doc/designs/passing-algorithmidentifier-parameters.md (hunk 2 above)
Comment (levitte) [reply doc/designs/passing-algorithmidentifier-parameters.md]: I've added some examples in the background section

**Comment by levitte** [comment]
@openssl/otc, please have a look

File: doc/designs/passing-algorithmidentifier-parameters.md [hunk 3]
```diff
@@ -0,0 +1,156 @@
+Passing AgorithmIdentifier parameters to operations
```
Comment (t8m) [inline doc/designs/passing-algorithmidentifier-parameters.md:1]: Typo AlgorithmIdentifier

File: doc/designs/passing-algorithmidentifier-parameters.md [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+[Background / tl;dr](#background-tldr) at the end of this design.
+
+Establish an OSSL_PARAM key that any algorithms may become aware of
+-------------------------------------------------------------------
+
+We already have a parameter key, but it's currently only specified for
+`EVP_CIPHER`, in support of `EVP_CIPHER_param_to_asn1()` and
+`EVP_CIPHER_asn1_to_param()`.
+
+"alg\_id\_param", also known as the macro `OSSL_CIPHER_PARAM_ALGORITHM_ID_PARAMS`
```
Comment (t8m) [inline doc/designs/passing-algorithmidentifier-parameters.md:25]: Why not use:
```
`"alg_id_param"`, also known...
```

File: doc/designs/passing-algorithmidentifier-parameters.md [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+- `doc/man7/provider-mac.pod`
+- `doc/man7/provider-signature.pod`
+
+That should cover all algorithms that are, or should be possible to fetch by
+AlgorithmIdentifier.algorithm, and for which there's potentially a relevant
+AlgorithmIdentifier.parameters field.
+
+We may arguably want to consider `doc/man7/provider-keymgmt.pod` too, but
+an AlgorithmIdentifier that's attached directly to a key is usually part of
+a PrivKeyInfo or SubjectÅublicKeyInfo structure, and those are handled by
```
Comment (t8m) [inline doc/designs/passing-algorithmidentifier-parameters.md:54]: Typo: SubjectPublicKeyInfo

File: doc/designs/passing-algorithmidentifier-parameters.md [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+endoders and decoders as those see fit, and there's no tangible reason why
```
Comment (t8m) [inline doc/designs/passing-algorithmidentifier-parameters.md:55]: Typo: encoders

**Review by t8m — CHANGES_REQUESTED** [review]
Just a few typos, otherwise LGTM

File: doc/designs/passing-algorithmidentifier-parameters.md (hunk 4 above)
Comment (levitte) [reply doc/designs/passing-algorithmidentifier-parameters.md:25]: 'cause some markdown interpreters interpret in-word underscores as italics markers too, so rendering that like this: alg*id*param .  See https://www.markdownguide.org/basic-syntax/#italic-best-practices for a note on the topic.
I keep forgetting how github markdown behaves...

**Review by t8m — APPROVED** [review]

**Review by mattcaswell — APPROVED** [review]

File: doc/designs/passing-algorithmidentifier-parameters.md [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+-------------------
+
+There are no possible fallback strategies, which is fine, considering that
+current provider functionality doesn't support passing AlgorithmIdentifier
+parameter data at all (except for `EVP_CIPHER`), and therefore do not work
+at all when such parameter data needs to be passed.
+
+-----
+
+-----
```
Comment (paulidale) [inline doc/designs/passing-algorithmidentifier-parameters.md:108]: ?????

Comment (levitte) [reply doc/designs/passing-algorithmidentifier-parameters.md:108]: Two horizontal lines separating the design and the background / tl;dr.  That a problem?

Comment (paulidale) [reply doc/designs/passing-algorithmidentifier-parameters.md:108]: Nope, just hadn't seen it before.

TL;DR usually comes at the top but I'm pretty flexible.

**Review by paulidale — APPROVED** [review]

File: doc/designs/passing-algorithmidentifier-parameters.md (hunk 7 above)
Comment (levitte) [reply doc/designs/passing-algorithmidentifier-parameters.md:108]: There's a link at the top

Comment (paulidale) [reply doc/designs/passing-algorithmidentifier-parameters.md:108]: Sure I know that but TL;DR **virtually always** gives a summery at the top rather than requiring a clickable link to the section below.

I suspect you've moved your head space to a more comfortable postfix mode rather than the more pure prefix version.

Comment (levitte) [reply doc/designs/passing-algorithmidentifier-parameters.md:108]: Okie.

**Comment by paulidale** [comment]
Merged.
